proxyutil/brw.go

```diff
@@ -0,0 +1,209 @@
+// buffered. If the response was compressed the body will be transparently
+// decompressed. Close must be called in order to clean up resources. GetBody
+// may be called multiple times, each time it will return an independent
+// ReadCloser for the original response body.
+func (brw *BufferedResponseWriter) GetBody() (io.ReadCloser, error) {
+	buf := brw.originalBodyRaw()
+	switch brw.Header().Get("Content-Encoding") {
+	case "gzip":
+		return gzip.NewReader(buf)
+	case "deflate":
+		return flate.NewReader(buf), nil
+	}
+	return ioutil.NopCloser(buf), nil
+}
+
+// SetBody sets the io.Reader from which the new body of the buffered response
+// should be read from. This io.Reader will be drained once ActuallyWrite is
+// called. If this is never called the original body will be used.
+//
+// Note: there is no need to send in compressed data here, the new response body
+// will be automatically compressed in the same way the original was.
+func (brw *BufferedResponseWriter) SetBody(in io.Reader) {
+	brw.newBody = in
+}
+
+// Returns a buffer containing the (potentially compressed) raw bytes which
+// comprise the body which should be written. If SetBody has been called then
+// that will be used as the body, otherwise the original bytes read in will be
+func (brw *BufferedResponseWriter) bodyBuf() (*bytes.Buffer, error) {
+	body := brw.originalBodyRaw()
+	if brw.newBody == nil {
+		return body, nil
+	}
+
+	// using body.Len() here is just a guess for how big we think the new body
+	// will be
+	bodyBuf := bytes.NewBuffer(make([]byte, 0, body.Len()))
+	dst := brw.bodyEncoder(bodyBuf)
+	defer dst.Close()
+
+	// dst will be some wrapper around bodyBuf, or bodyBuf itself
+	if _, err := io.Copy(dst, brw.newBody); err != nil {
+		return nil, err
+	}
+	return bodyBuf, nil
+}
+
+// ActuallyWrite takes all the buffered data and actually writes to the wrapped
+// ResponseWriter. Returns the number of bytes written as the body
+func (brw *BufferedResponseWriter) ActuallyWrite() (int64, error) {
+	bodyBuf, err := brw.bodyBuf()
+	if err != nil {
+		return 0, err
+	}
+
+	rw := brw.ResponseWriter
+	rw.Header().Set("Content-Length", strconv.Itoa(bodyBuf.Len()))
+	rw.WriteHeader(brw.code)
+	return io.Copy(rw, bodyBuf)
+}
+
+type nopWriteCloser struct {
+	io.Writer
+}
+
+func (n nopWriteCloser) Close() error {
+	if wc, ok := n.Writer.(io.WriteCloser); ok {
+		return wc.Close()
+	}
+	return nil
+}
+
+func (brw *BufferedResponseWriter) bodyEncoder(w io.Writer) io.WriteCloser {
+	switch brw.Header().Get("Content-Encoding") {
+	case "gzip":
+		return gzip.NewWriter(w)
+	case "deflate":
+		// From the docs: If level is in the range [-1, 9] then the error
+		// returned will be nil. Otherwise the error returned will be non-nil
+		// so we can ignore error
+		dfW, _ := flate.NewWriter(w, -1)
+		return dfW
+	default:
+		return nopWriteCloser{w}
+	}
+}
+
+type brwMarshalled struct {
+	Code   int
+	Header map[string][]string
+	Body   []byte
+}
+
+var _ = func() bool {
```

Comment:
i guess now I can. When this was in rpcutil it was just kinda weird to have it so "far away" in the code